src/pieces.py

```diff
@@ -9,118 +9,272 @@
 
         Returns:
-            Whether the piece is a black piece or not.
+            A list of all possible moves.
         """
-        return self.color == "black"
+        return set()  # A ghost piece cannot move.
 
 
-@dataclass
 class Pawn(Piece):
-    """A pawn."""
+    """A pawn.
+
+    Moves forward a square, unless its first move where it can move two and unless blocked by any piece.
+    It captured forward-diagonally. I can capture en-passant:
+        If an enemy pawn passes yours trying to escape capture, well... it cannot, not for this round.
+    NOTE: That probably means that board will invoke this method multiple times (unti a better way is thought).
+    Upon reaching the end of the board, it is promoted to an officer:
+        - Rook
+        - Bishop
+        - Knight
+        - Queen
+    """
+
+    pawn_step = Piece.south
+
+    steps = Piece.steps.union({
+        pawn_step,  # One step.
+        pawn_step + Piece.west,  # Capturing to the left.
+        pawn_step + Piece.east,  # Capturing to the right.
+    })
 
     value: int = 1
 
-    def __repr__(self) -> str:
-        """Represent the pawn.
+#   Extra flags for the pawn:
+    has_moved: bool = False
 
-        Returns:
-            The representation of a pawn.
-        """
+    def __repr__(self) -> str:
+        self.__repr__.__doc__
         return "♟" if self.is_black else "♙"
 
+#   NOTE: Probably a board will invoke this method multiple times (until a better way is thought).
+    def legal_moves(self, square: Square, condition: Callable[[Square], bool]) -> set[Square]:
+        super().legal_moves.__doc__
+        squares = set()
+        for move in self.steps:
+            if condition(square + move * self.color.value):
+                squares.add(square + move * self.color.value)
+        if not self.has_moved:
+            squares.add(square + (self.pawn_step + self.pawn_step) * self.color.value)
+        return squares
+
 
 @dataclass
-class Bishop(Piece):
-    """A bishop."""
+class Melee(Piece):
+    """A close ranged piece:
+        - King
+        - Knight (has reach)
 
-    value: int = 3
+    Pawns are special.
+    """
 
-    def __repr__(self) -> str:
-        """Represent the bishop.
+    def legal_moves(self, square: Square, condition: Callable[[Square], bool]) -> set[Square]:
+        super().legal_moves.__doc__
+        squares = set()
+        for move in self.steps:
+            if condition(square + move):
+                squares.add(square + move)
+        return squares
 
-        Returns:
-            The representation of a bishop.
-        """
-        return "♝" if self.is_black else "♗"
+
+@dataclass
+class King(Melee):
+    """A King.
+
+    Moves one square in any direction that is not blocked by same side piece or targeted by enemy piece.
+    """
+
+    steps = Melee.steps.union(Piece.straights.union(Piece.diagonals))
```

Comment:
```py
steps = set.union(Piece.straights, Piece.diagonals)
```

Reply:
Hmm, doesn't update comments in the review tab, that's why I didn't see your comment.